c10/core/ModePyObjTrampoline.h

```diff
@@ -0,0 +1,32 @@
+#pragma once
+
+#include <c10/core/SafePyObject.h>
+
+namespace c10 {
+
+// this is basically just a light wrapper around a SafePyObject that also has
+// some way of declaring a push and pop trampoline. This lets us write those
+// methods in the torch namespace where we have access to pybind
```

Comment:
Basic design of this version looks like:
every time that we push, pop, or copy the stack, we call a python function on the mode that checks the state and updates the boolean on the mode that says whether it's on the stack or not

There's another version of this that @zou3519 and I talked through: the mode stack is actually vector and a set together in a data structure. We only allow push/pop operations on this data structure and it's responsible of keeping track of what's around

Summary of benefits/costs to each
**Pros of C++ data structure (set + vector)**
- easier to understand
- all logic for mode stack lives in C++
  -  no jumps back to Python

**Pros of Python function version (this version)**
- don't have repetition of the same information (vector + set together has code smell, even if there's safety because of the functions we expose)
- user can decide what it means to be on the stack by overriding the `check_mode_state_*` functions
  - right now the default is that two modes of the same Python type can be on the stack but the exact same mode can't be on the stack. Fake Tensor will want to restrict such that no two Fake Tensor modes can exist on the stack at the same time

Any thoughts/preferences welcome!

Reply:
Yeah I feel like this is the big question. Let me offer one more possibility: handling ALL of the state in Python. The reason we moved it to C++ in the first place was because the TLS save load logic only knew how to deal with C++ objects. If we have a way of registering arbitrary Python to be saved and restored, there isn't a reason outside of efficiency to do it in C++

Reply:
Yeah I feel like this is the big question. Let me offer one more possibility: handling ALL of the state in Python. The reason we moved it to C++ in the first place was because the TLS save load logic only knew how to deal with C++ objects. If we have a way of registering arbitrary Python to be saved and restored, there isn't a reason outside of efficiency to do it in C++